### PR TITLE
Add PyProj to packages

### PR DESCRIPTION
**Pull request by dopplershift**
### Description

This adds PyProj to the packages, as well as its dependency PROJ (libproj). This required pointing directly to the sqlite3.wasm that's built with Python (or else the build refused to detect the dependency as present).

I also tried to build as a shared library, but it said that wasn't supported. 🤷‍♂️ 

### Checklists

- [ ] Add a [CHANGELOG](https://github.com/pyodide/pyodide/blob/main/docs/project/changelog.md) entry
- [ ] Add / update tests
- [ ] Add new / update outdated documentation

### CODE REVIEW

**Comment by dopplershift** [comment]
The build is failing with:
```
-- Configuring incomplete, errors occurred!
See also "/root/repo/packages/libproj/build/libproj-8.2.1/build/CMakeFiles/CMakeOutput.log".
emcmake: error: 'cmake ../ -DCMAKE_INSTALL_PREFIX=../../proj -DCMAKE_BUILD_TYPE=Release -DENABLE_CURL=OFF -DENABLE_TIFF=OFF -DBUILD_APPS=OFF -DBUILD_SHARED_LIBS=OFF -DBUILD_TESTING=OFF -DSQLITE3_LIBRARY=/root/repo/cpython/build/sqlite-autoconf-3380500/sqlite3.wasm -DSQLITE3_INCLUDE_DIR=/root/repo/cpython/build/sqlite-autoconf-3380500/ -DCMAKE_C_FLAGS=-fPIC -DCMAKE_CXX_FLAGS=-fPIC -DCMAKE_TOOLCHAIN_FILE=/root/repo/emsdk/emsdk/upstream/emscripten/cmake/Modules/Platform/Emscripten.cmake -DCMAKE_CROSSCOMPILING_EMULATOR=/root/repo/emsdk/emsdk/node/14.18.2_64bit/bin/node' failed (returned 1)
[2022-05-13 17:12:51] Building package libproj...
ERROR: bash command failed
    mkdir -p build;
    cd build \
        && LDFLAGS="-s NODERAWFS=1 -sUSE_ZLIB=1 -sFORCE_FILESYSTEM=1" emcmake cmake ../ \
        -DCMAKE_INSTALL_PREFIX=../../proj \
        -DCMAKE_BUILD_TYPE=Release \
        -DENABLE_CURL=OFF \
        -DENABLE_TIFF=OFF \
        -DBUILD_APPS=OFF \
        -DBUILD_SHARED_LIBS=OFF \
        -DBUILD_TESTING=OFF \
        -DSQLITE3_LIBRARY=$PYODIDE_ROOT/cpython/build/sqlite-autoconf-3380500/sqlite3.wasm \
        -DSQLITE3_INCLUDE_DIR=$PYODIDE_ROOT/cpython/build/sqlite-autoconf-3380500/ \
        -DCMAKE_C_FLAGS="-fPIC" \
        -DCMAKE_CXX_FLAGS="-fPIC";
    emmake make -j ${PYODIDE_JOBS:-3} install;

[2022-05-13 17:13:03] Succeeded building package libproj in 13.0 seconds.
ERROR: cancelling buildall
Fatal Python error: _enter_buffered_busy: could not acquire lock for <_io.BufferedWriter name='<stdout>'> at interpreter shutdown, possibly due to daemon threads
Python runtime state: finalizing (tstate=0x0000559dad6cdc10)

Current thread 0x00007efc1ca5e740 (most recent call first):
  <no Python frame>

Extension modules: _ruamel_yaml, yaml._yaml (total: 2)
make: *** [Makefile:13: all] Aborted (core dumped)
make: Leaving directory '/root/repo/packages'

Exited with code exit status 
```
and I have no idea what that would mean...

**Comment by ryanking13** [comment]
That error message is a general error message for the build failure.

I guess you'll need to look at `build.log` of `libproj` or `CMakeOutput.log` as the error message says.

**Comment by hoodmane** [comment]
You should change the CI script to add an extra `store_artifacts` step that stores `packages/libproj` as an artifact:
https://github.com/pyodide/pyodide/blob/dc29ae65416e624e1fc2bc3b48c65f71a08ce99a/.circleci/config.yml#L166

**Comment by ryanking13** [comment]
Actually, `build.log` is printed to the stdout if the build fails, so you can see the CI log instead.

It looks like there is an issue with sqlite3

```
CMake Error at CMakeLists.txt:181 (message):
  sqlite3 binary not found!


CMake Error at cmake/FindSqlite3.cmake:59 (file):
  file STRINGS file
  "/root/repo/cpython/build/sqlite-autoconf-3380500/sqlite3.h" cannot be
  read.
Call Stack (most recent call first):
  CMakeLists.txt:184 (find_package)


-- Found Sqlite3: /root/repo/cpython/build/sqlite-autoconf-3380500/sqlite3.wasm
-- Sqlite3 version: 
CMake Error at CMakeLists.txt:192 (message):
  sqlite3 >= 3.11 required!
```

File: packages/libproj/meta.yaml
Comment (ryanking13) [inline packages/libproj/meta.yaml:22]: I think this should point to a `.a` file.

```suggestion
        -DSQLITE3_LIBRARY=$PYODIDE_ROOT/cpython/build/sqlite-autoconf-3380500/.libs/libsqlite3.a \
```

**Comment by ryanking13** [comment]
BTW, @hoodmane, if some packages require libsqlite3 as a dependency do you think we should move it to packages similar to `openssl` and `ssl`?

**Comment by hoodmane** [comment]
Well it depends on whether we think `sqlite` should always be loaded. But I think ideally we would build `sqlite` as a shared library, only load it on request via `pyodide.loadPackage`, and link it dynamically into other dependent packages.

File: packages/libproj/meta.yaml
Comment (dopplershift) [reply packages/libproj/meta.yaml:22]: Ah, I didn't think to check `.libs`.

**Comment by dopplershift** [comment]
Ok, I'm not understanding:
```
CMake Error at cmake/FindSqlite3.cmake:59 (file):
  file STRINGS file
  "/root/repo/cpython/build/sqlite-autoconf-3380500/sqlite3.h" cannot be
  read.
Call Stack (most recent call first):
  CMakeLists.txt:184 (find_package)
```
the file is there on my system? How does it have `.libs` but not the include?

**Comment by ryanking13** [comment]
It looks like we remove CPython build artifacts after the build in CI so you can't use sqlite3 library (abd also its headers) when building packages.

https://github.com/pyodide/pyodide/blob/eb4f7d9fd9b5d870ac72783334666a99cd2afe4e/.circleci/config.yml#L90-L94

I think you'll need to add sqlite3 in `packages` just like other libraries.

**Comment by dopplershift** [comment]
Ok, added sqlite3 as a separate package. That seems to be found.

PROJ needs an sqlite3 executable in order to build its internal database. I tried adding this to the `Dockerfile`, but it doesn't seem to have worked. Also, not sure if that's how y'all want to see that accomplished.

**Comment by ryanking13** [comment]
> Ok, added sqlite3 as a separate package. That seems to be found.
> 
> PROJ needs an sqlite3 executable in order to build its internal database. I tried adding this to the `Dockerfile`, but it doesn't seem to have worked. Also, not sure if that's how y'all want to see that accomplished.

Can you try downloading sqlite3 binary during the build (adding a build script section in meta.yaml)? I think we prefer that way than installing it globally.

for example, you can do something like (it is how I did when we need protoc binary during the build):

```yaml
build:
  script: |
    # Note: We need protoc binary during the build
    case "$(uname -s)" in
        Linux*) export MACHINE=linux;;
        Darwin*) export MACHINE=osx;;
        *) export MACHINE="UNKNOWN"
    esac
    
    export PROTOC_VERSION="3.13.0"
    export PROTOC_ZIP="protoc-${PROTOC_VERSION}-${MACHINE}-x86_64.zip"
    export PROTOC_URL="https://github.com/protocolbuffers/protobuf/releases/download/v${PROTOC_VERSION}/${PROTOC_ZIP}"
    
    wget ${PROTOC_URL}
    unzip ${PROTOC_ZIP} -d protoc
    export PATH=${PATH}:$(pwd)/protoc/bin/
```

**Comment by ryanking13** [comment]
> > Ok, added sqlite3 as a separate package. That seems to be found.
> > PROJ needs an sqlite3 executable in order to build its internal database. I tried adding this to the `Dockerfile`, but it doesn't seem to have worked. Also, not sure if that's how y'all want to see that accomplished.
> 
> Can you try downloading sqlite3 binary during the build (adding a build script section in meta.yaml)? I think we prefer that way than installing it globally.
> 
> for example, you can do something like (it is how I did when we need protoc binary during the build):
> 
> ```yaml
> build:
>   script: |
>     # Note: We need protoc binary during the build
>     case "$(uname -s)" in
>         Linux*) export MACHINE=linux;;
>         Darwin*) export MACHINE=osx;;
>         *) export MACHINE="UNKNOWN"
>     esac
>     
>     export PROTOC_VERSION="3.13.0"
>     export PROTOC_ZIP="protoc-${PROTOC_VERSION}-${MACHINE}-x86_64.zip"
>     export PROTOC_URL="https://github.com/protocolbuffers/protobuf/releases/download/v${PROTOC_VERSION}/${PROTOC_ZIP}"
>     
>     wget ${PROTOC_URL}
>     unzip ${PROTOC_ZIP} -d protoc
>     export PATH=${PATH}:$(pwd)/protoc/bin/
> ```

File: packages/libproj/meta.yaml
Comment (ryanking13) [inline packages/libproj/meta.yaml:16]:
```suggestion
```

File: packages/libproj/meta.yaml
Comment (ryanking13) [inline packages/libproj/meta.yaml:28]: It looks like there is a subdirectory inside a sqlite zip file (maybe this is wrong, please check by yourself):

```suggestion
    export PATH=${PATH}:$(pwd)/sqlite/sqlite-tools-${MACHINE}-x86-${SQLITE_VERSION}
```

File: packages/libproj/meta.yaml
Comment (ryanking13) [inline packages/libproj/meta.yaml:11]: I think you need to add libtiff here:

```suggestion
    - sqlite3
    - libtiff
```

Comment (dopplershift) [reply packages/libproj/meta.yaml:11]: Thanks, that might fix libtiff, but I still need to figure out why the pre-built sqlite3 is failing when the build tries to use it.

Comment (ryanking13) [reply packages/libproj/meta.yaml:11]: > Thanks, that might fix libtiff, but I still need to figure out why the pre-built sqlite3 is failing when the build tries to use it.

https://stackoverflow.com/questions/14082152/sqlite3-no-such-file-or-directory

Hmm it seems like pre-built sqlite3 binary doesn't work for 64bit systems. I'm not sure why they don't provide x86-64 binaries. I think we will have to build sqlite3 for host os or add sqlite3 binary in our docker image as you mentioned before...

Comment (dopplershift) [reply packages/libproj/meta.yaml:11]: Including a `sqlite3` binary like you would a compiler (using a distro package) seems like the easiest path here...

Comment (ryanking13) [reply packages/libproj/meta.yaml:11]: Alright, let's add sqlite3.

Comment (dopplershift) [reply packages/libproj/meta.yaml:11]: Can I just add it to the `Dockerfile`? Or does it need a separate PR?

Comment (ryanking13) [reply packages/libproj/meta.yaml:11]: I'll open a PR. One of core devs need to trigger the Docker deploy workflow.

**Comment by ryanking13** [comment]
I opened a PR adding sqlite3 in our docker image (#2807). You can try it with: `pyodide/pyodide-env:20220629-py310-chrome102-firefox100`.

**Comment by dopplershift** [comment]
🤞

**Comment by ryanking13** [comment]
It looks like now it wants `PROJ_DIR` env variable to be set?

**Comment by dopplershift** [comment]
Has anyone tried making a `certifi` package for inclusion in the pyodide runtime?

File: packages/pyproj/patches/fix-build.patch
Comment (ryanking13) [inline packages/pyproj/patches/fix-build.patch:9]: Perhaps we should update `pywasmcross.py` to ignore `-R` flags in `wasm-ld` (no action required for this PR) @hoodmane.

File: packages/certifi/meta.yaml
Comment (ryanking13) [inline packages/certifi/meta.yaml:6]: certifi has a pure python wheel. Could you replace the source from `.tar.gz` with `.whl`?

**Review by ryanking13 — COMMENTED** [review]
Thanks, @dopplershift! It looks very cool and I hope you had some fun working on this. Now you'll need to:

1. Add some basic tests for pyproj
2. Update changelog
3. Remove sqlite tar file that I think you accidentally pushed
4. Wait until #2807 gets merged and merge main to this branch

File: packages/certifi/meta.yaml
Comment (dopplershift) [reply packages/certifi/meta.yaml]: Done.

File: packages/pyproj/patches/fix-build.patch
Comment (dopplershift) [reply packages/pyproj/patches/fix-build.patch:9]: This would save me upstreaming the patch.

**Comment by dopplershift** [comment]
> Thanks, @dopplershift! It looks very cool and I hope you had some fun working on this. Now you'll need to:
> 
> 1. Add some basic tests for pyproj

Can you clarify what it would look like? I checked h5py, but didn't see anything.

> 2. Update changelog

Done.

> 3. Remove sqlite tar file that I think you accidentally pushed

Done.

> 4. Wait until #2807 gets merged and merge main to this branch

Ok.

**Comment by ryanking13** [comment]
> > 1. Add some basic tests for pyproj
> 
> Can you clarify what it would look like? I checked h5py, but didn't see anything.

You need to add `test_pyproj.py` inside pyproj directory.

See `pyyaml` example for a reference: https://github.com/pyodide/pyodide/blob/main/packages/pyyaml/test_pyyaml.py

**Comment by dopplershift** [comment]
Added the test and rebased on main with #2807.

**Comment by dopplershift** [comment]
Good to have the test because it caught:
```
/lib/python3.10/site-packages/pyproj/__init__.py:91: UserWarning: Valid PROJ data directory not found. Either set the path using the environmental variable PROJ_LIB or with `pyproj.datadir.set_data_dir`.
  warnings.warn(str(err))
Pyodide has suffered a fatal error. Please report this to the Pyodide maintainers.
The cause of the fatal error was:
CppException osgeo::proj::io::FactoryException: Cannot find proj.db
```
Where are the data files supposed to be in our environment? What's the best way to point to them? Environment variable? Edit the package files?

**Comment by ryanking13** [comment]
> Apparently it can get included in the wheel for `pyproj` if you set `PROJ_WHEEL` and have a copy of PROJ in a specific internal directory. So I guess to be more clear, the file `proj.db` and a few other data files should be part of the PROJ C library package (that I added here as `libproj`).

I added a script that copies proj.db and data files into pyproj directory, following the setup.py of pyproj. 

https://github.com/pyproj4/pyproj/blob/54ad87816a239ad7f42dd626beb18341987b01f0/setup.py#L199-L214

Probably in the future it would be good to ask the pyproj maintainer to add a feature to modify the search path by using env variables.

File: packages/pyproj/meta.yaml
Comment (ocefpaf) [inline packages/pyproj/meta.yaml:24]: I'm guessing this is the best option for now but that would duplicate all these files in projects that depend both on proj and pyproj, right?

Comment (ryanking13) [reply packages/pyproj/meta.yaml:24]: Probably, yes. I don't know much about libproj, but it seems like packages that use libproj all embeds `pyproj.db` inside its wheel. For example, inside [fiona](https://pypi.org/project/Fiona/), I can see `pyproj.db`.

```
$ unzip -l Fiona-1.8.21-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl  | grep proj.db
  7921664  2022-02-07 15:46   fiona/proj_data/proj.db
```

Since `proj.db` is a big file in terms of the browser (~7.7MB), it would be great if we can share it between multiple packages. Maybe someone can ask package maintainers to do such optimization.

**Review by ryanking13 — APPROVED** [review]
I think we can merge this now, thanks @dopplershift for your hard work!

**Comment by dopplershift** [comment]
🎉 

@ryanking13 So just to be clear, it gets bundled in wheels because they need to bundle all of libproj because wheels/PyPI don't have a concept of non-Python dependencies. On the Conda side, because it just handles generic packages, the libproj package has `proj.db` that everything else (like fiona and pyproj) can "just use".

It's unclear to me which approach packaging within pyodide is taking.

File: packages/pyproj/meta.yaml
Comment (ocefpaf) [reply packages/pyproj/meta.yaml:24]: It is a design problem with the wheel. They a need to bundle proj and its data and we ended up with multiple copies of it. Package manager that have a more "ecosystem" logic can share data and c-lib dependencies in a more efficient way.

I'm not a pyodide expert but maybe we can use the same approach here as we do with conda?

Comment (ryanking13) [reply packages/pyproj/meta.yaml:24]: I think we will keep using wheels for compatibility, but that sound very interesting, thanks.

**Pull request merged**
